### PR TITLE
feat(react-tabs): export Tab component style hooks

### LINKED ISSUE #32078: [Feature]: Add InteractiveTab component to react-tabs
### Library

React Components / v9 (@fluentui/react-components)

### Describe the feature that you would like added

TabList should be able to support interactive tabs. For example, a tab with a dismiss/close button that removes the tab or a tab with a dropdown or searchbox on it.

The root slot for Tab is button with role=tab.  Aria specifies that anything with role=tab have presentational children only. Also a button containing a button is not valid HTML for browsers.

To solve this, we need InteractiveTab.  It would be based on Tab, but have the root be a div. It should have 3 additional slots: button (where role=tab is applied), beforeContent, and afterContent. Similar to the Input component this would allow callers to place interactive content before or after the tab button.

The Tab styles hook should have subhooks to appy the tab button style vs. the indicator styles.  InteractiveTab would apply the indicator styles to the root div and the button styles to the button slot.  The Tab style hook would call these internally to apply both styles to the root button so there would be no breaking change to Tab.

### Have you discussed this feature with our team

Tudor

### Additional context

Partners are in progress trying to recompose TabList and Tab to have this feature and struggling to not cut-and-paste everything from react-tabs. By adding InteractiveTab, we alleviate the divergence and support a common scenario.

### Validations

- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Priority

High

### PR DESCRIPTION
**Pull request by dmytrokirpa**
## Previous Behavior

Previously, TabList did not support interactive tabs, such as a tab with a dismiss/close button to remove the tab, or a tab featuring a dropdown or search box.

The primary slot for a Tab was a button with `role=tab`. According to Aria guidelines, elements with `role=tab` should only contain presentational children.

## New Behavior

This PR introduces the `InteractiveTab` component, which is derived from `Tab`, but instead uses a div as its root. It includes three additional slots: `button` (where `role=tab` is assigned), `beforeContent`, and `afterContent`. This design, akin to the Input component, allows users to insert interactive content before or after the tab button.

The styling hook for Tab has been segmented into sub-hooks to differentiate the application of tab button styles from indicator styles. The `InteractiveTab` assigns indicator styles to the root div and button styles to the `button` slot. The Tab's styling hook utilizes these sub-hooks internally to apply both styles to the root button, ensuring no breaking changes to the Tab.

## Related Issue(s)

- Fixes #32078

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

File: packages/react-components/react-components/src/index.ts [hunk 1]
```diff
@@ -722,6 +722,7 @@ export {
 } from '@fluentui/react-switch';
```
Comment (fabricteam) [inline packages/react-components/react-components/src/index.ts:722]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=32125&commitId=d157f7a71e6ca05cac047e5748f67299995d3885&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask.normal.chromium.png|5| Changed|
|Avatar Converged.badgeMask - RTL.normal.chromium.png|2| Changed|

</details>


  <details><summary><b>Card Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged.appearance - High Contrast.normal.chromium.png|0| Added|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Avatar-mount_base.html">671</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Avatar-mount.html">632</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Button-mount_base.html">306</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Button-mount.html">306</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Field-mount_base.html">1145</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Field-mount.html">1152</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProvider-mount_base.html">718</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProvider-mount.html">698</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProviderWithTheme-mount_base.html">82</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProviderWithTheme-mount.html">97</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">34</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">43</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">72</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">82</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/MakeStyles-mount_base.html">872</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/MakeStyles-mount.html">849</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Persona-mount_base.html">1809</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/Persona-mount.html">1789</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/SpinButton-mount_base.html">1401</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/SpinButton-mount.html">1436</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/SwatchPicker-mount_base.html">1673</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32125/perf-test-react-components/SwatchPicker-mount.html">1709</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: change/@fluentui-react-components-7da82804-2b64-42a2-a91e-042686906623.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-components-7da82804-2b64-42a2-a91e-042686906623.json:1]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=32125&commitId=1a7b11d09376ecb68d625a3cebcae47b9e80c29b&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask - RTL.normal.chromium.png|5| Changed|
|Avatar Converged.badgeMask.normal.chromium.png|5| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-shared-contexts/library/src/CustomStyleHooksContext/CustomStyleHooksContext.ts [hunk 3]
```diff
@@ -62,6 +62,7 @@ export type CustomStyleHooksContextValue = Partial<{
   useSpinnerStyles_unstable: CustomStyleHook;
   useSwitchStyles_unstable: CustomStyleHook;
   useTabStyles_unstable: CustomStyleHook;
+  useInteractiveTabStyles_unstable: CustomStyleHook;
```
Comment (GeoffCoxMSFT) [inline packages/react-components/react-shared-contexts/library/src/CustomStyleHooksContext/CustomStyleHooksContext.ts:65]: It is awesome that you thought to add this!  I would hold off on adding it to the type until the component is out of preview and public.

File: packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  /**
+   * Root of the component.
+   */
+  root: Slot<'div'>;
+
+  /**
+   * button
+   */
+  button: NonNullable<TabSlots['root']>;
+
```
Comment (GeoffCoxMSFT) [inline packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts:14]: Consider adding comments to these slots.

File: packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import type { TabSlots, TabInternalProps, TabInternalState } from '../Tab/Tab.types';
+
+export type InteractiveTabSlots = {
+  /**
+   * Root of the component.
+   */
+  root: Slot<'div'>;
+
+  /**
+   * button
```
Comment (GeoffCoxMSFT) [inline packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts:11]: Consider expanding this comment to be more meaningful.

File: packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+  contentAfter?: Slot<'span'>;
+} & Omit<TabSlots, 'root'>;
+
+export type InteractiveTabProps = Omit<ComponentProps<Partial<InteractiveTabSlots>, 'button'>, 'content'> &
+  Pick<Partial<InteractiveTabSlots>, 'content'> &
+  TabInternalProps;
+
+export type InteractiveTabInternalSlots = InteractiveTabSlots & {
+  contentReservedSpace?: Slot<'span'>;
```
Comment (GeoffCoxMSFT) [inline packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts:25]: Consider intersecting with TabInternalSlots

File: packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+   * button
+   */
+  button: NonNullable<TabSlots['root']>;
+
+  contentBefore?: Slot<'span'>;
+
+  contentAfter?: Slot<'span'>;
+} & Omit<TabSlots, 'root'>;
+
+export type InteractiveTabProps = Omit<ComponentProps<Partial<InteractiveTabSlots>, 'button'>, 'content'> &
```
Comment (GeoffCoxMSFT) [inline packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts:20]: I don't understanding omitting button and content, then adding back content.

Bernardo changed the TabProps back in January to something that doesn't make sense to me either:

export type TabProps = Omit<ComponentProps<Partial<TabSlots>>, 'content'> &
  Pick<Partial<TabSlots>, 'content'> &

Can you chat with him about what is going on here?

**Review by GeoffCoxMSFT — APPROVED** [review]

**Review by GeoffCoxMSFT — APPROVED** [review]

File: packages/react-components/react-shared-contexts/library/src/CustomStyleHooksContext/CustomStyleHooksContext.ts (hunk 3 above)
Comment (dmytrokirpa) [reply packages/react-components/react-shared-contexts/library/src/CustomStyleHooksContext/CustomStyleHooksContext.ts]: Done, thanks for the feedback!

File: packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts (hunk 5 above)
Comment (dmytrokirpa) [reply packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts]: Yes, that was a placeholder for a comment. I will tidy it up before submitting it for review.

File: packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts (hunk 7 above)
Comment (dmytrokirpa) [reply packages/react-components/react-tabs/library/src/components/InteractiveTab/InteractiveTab.types.ts]: It was copied over from the Tab component, I simplified the type

**Comment by Hotell** [comment]
I'll check the actionables that we discussed offline last week tomorrow - sorry for delay

**Review by Hotell — COMMENTED** [review]
ok so please correct me if I understood the code gymnastics incorrectly.

from what I see, in order to ship this as extension we need following shared APIs to export from react-tabs public api surface:

types:
- TabInternalProps
- TabInternalState

styles:
- useTabIndicatorStyles
- useTabButtonStyles
- useTabContentStyles

_For types:_

we don't need to expose these things and do the Type transform within extension from already exposed api, eg: `Pick<TabProps,...>` same for the `TabState`

_For style:_

It seems those atomic hooks might be a good idea to have exposed anyways for any kind of modification, so IMO it can be exported from `react-tabs` but I'd hold on re-exporting those from `react-components` suite.

with this approach we can build this control as extension 
- using imports from `react-tabs` directly ( for the styles)
- using imports from `react-components` suite ( for the already exposed types ) and do transforms within extensions code  


Thoughts? ty

**Comment by dmytrokirpa** [comment]
> ok so please correct me if I understood the code gymnastics incorrectly.
> 
> from what I see, in order to ship this as extension we need following shared APIs to export from react-tabs public api surface:
> 
> types:
> 
> * TabInternalProps
> * TabInternalState
> 
> styles:
> 
> * useTabIndicatorStyles
> * useTabButtonStyles
> * useTabContentStyles
> 
> _For types:_
> 
> we don't need to expose these things and do the Type transform within extension from already exposed api, eg: `Pick<TabProps,...>` same for the `TabState`
> 
> _For style:_
> 
> It seems those atomic hooks might be a good idea to have exposed anyways for any kind of modification, so IMO it can be exported from `react-tabs` but I'd hold on re-exporting those from `react-components` suite.
> 
> with this approach we can build this control as extension
> 
> * using imports from `react-tabs` directly ( for the styles)
> * using imports from `react-components` suite ( for the already exposed types ) and do transforms within extensions code
> 
> Thoughts? ty

Absolutely agree, I have updated the PR to reflect those changes.

File: packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts [hunk 8]
```diff
@@ -440,27 +447,28 @@ const useContentStyles = makeStyles({
...
+  useTabIndicatorStyles_unstable(state);
+
+  useTabButtonStyles_unstable(state, state.root);
+
+  useTabContentStyles_unstable(state);
+
+  return state;
+};
+
+/** Apply styling to the Tab indicator slot based on the state */
```
Comment (Hotell) [inline packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts:459]: nit: lets use jsdoc in consistent fashion

```suggestion
/** 
 * Apply styling to the Tab indicator slot based on the state 
 */
```

File: packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts [hunk 9]
```diff
@@ -486,6 +494,47 @@ export const useTabStyles_unstable = (state: TabState): TabState => {
     state.root.className,
   );
 
+  useTabAnimatedIndicatorStyles_unstable(state);
+
+  return state;
+};
+
+/** Apply styling to the Tab button slot based on the state */
```
Comment (Hotell) [inline packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts:502]: https://github.com/microsoft/fluentui/pull/32125/files#r1725345146

File: packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts [hunk 10]
```diff
@@ -486,6 +494,47 @@ export const useTabStyles_unstable = (state: TabState): TabState => {
...
+    !disabled && selected && rootStyles.selected,
+    disabled && rootStyles.disabled,
+
+    slot.className,
+  );
+
+  return state;
+};
+
+/** Apply styling to the Tab content slot based on the state */
```
Comment (Hotell) [inline packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts:529]: https://github.com/microsoft/fluentui/pull/32125/files#r1725345146

File: packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts (hunk 8 above)
Comment (Hotell) [inline packages/react-components/react-tabs/library/src/components/Tab/useTabStyles.styles.ts:459]: can we add more verbose docs about these new low level style hook apis ?

something like how/why would one use these including mentioning that these are part off/called  from main  `useTabStyles_unstable` so for composition this should be used only in very rare cases etc ?

**Review by Hotell — APPROVED** [review]

**Pull request merged**